developer/src/tike/project/Keyman.Developer.UI.Project.xmlLdmlProjectFileUI.pas

```diff
@@ -0,0 +1,285 @@
+
+interface
+
+uses
+  System.UITypes,
+  Menus,
+  Keyman.Developer.UI.Project.ProjectFilesUI,
+  Keyman.Developer.UI.Project.ProjectFileUI,
+  Keyman.Developer.UI.Project.ProjectUIFileType,
+  UfrmMessages,
+  Keyman.Developer.System.Project.xmlLdmlProjectFile,
+  Keyman.Developer.System.Project.xmlLdmlProjectFileAction;
+
+type
+  TxmlLdmlProjectFileUI = class(TOpenableProjectFileUI)
+  private
+    function TestKeymanWeb(FSilent: Boolean): Boolean;
+    function DebugKeyboard(FSilent: Boolean): Boolean;
+    function InstallKeyboard: Boolean;
+    function UninstallKeyboard: Boolean;
+    function GetProjectFile: TxmlLdmlProjectFileAction;
+
+    function GetDebug: Boolean;
+    procedure SetDebug(const Value: Boolean);
+    function CompileKeyboard(FSilent: Boolean): Boolean;
+    function TestKeyboardState(FCompiledName: string; FSilent: Boolean): Boolean;
+  public
+    function DoAction(action: TProjectFileAction; FSilent: Boolean): Boolean; override;
+    property Debug: Boolean read GetDebug write SetDebug;
+    property ProjectFile: TxmlLdmlProjectFileAction read GetProjectFile;
+  end;
+
+implementation
+
+uses
+  Winapi.Windows,
+  System.Classes,
+  System.StrUtils,
+  System.SysUtils,
+  System.Variants,
+  Vcl.Dialogs,
+  Vcl.Graphics,
+  Vcl.Controls,
+  dmActionsMain,
+
+  Keyman.Developer.UI.UfrmMessageDlgWithSave,
+  UfrmMain,
+  UfrmEditor,
+  UfrmMDIEditor,
+  UKeymanTargets,
+  UmodWebHttpServer,
+  Keyman.Developer.System.ServerAPI,
+  Keyman.Developer.UI.ServerUI,
+  KeyboardFonts,
+  KeymanDeveloperUtils,
+  KeymanDeveloperOptions,
+  utilsystem;
+
+function TxmlLdmlProjectFileUI.DoAction(action: TProjectFileAction; FSilent: Boolean): Boolean;
+begin
+  case action of
+    pfaCompile:   Result := CompileKeyboard(FSilent);
+    pfaTestKeymanWeb: Result := TestKeymanWeb(FSilent);
+    pfaInstall:   Result := InstallKeyboard;
+    pfaUninstall: Result := UninstallKeyboard;
+//  pfaDebug:     Result := DebugKeyboard(FSilent);
+    pfaClean:      Result := ProjectFile.Clean;
+  else
+    Result := False;
+  end;
+end;
+
+function TxmlLdmlProjectFileUI.CompileKeyboard(FSilent: Boolean): Boolean;
+var
+  FSave: Boolean;
+begin
+  Result := False;
+
+  if ProjectFile.Modified then
+  begin
+    if not FSilent then
+    begin
+      if not FKeymanDeveloperOptions.AutoSaveBeforeCompiling then
+      begin
+        if TfrmMessageDlgWithSave.Execute(
+            'The keyboard file has been modified.  You must save before compiling.'+#13#10#13#10+
+            'Save the keyboard and continue?',
+            'Always save automatically before compiling',
+            '', True, FSave) in [mrNo, mrCancel] then
+          Exit(False);
+        if FSave then
+        begin
+          FKeymanDeveloperOptions.AutoSaveBeforeCompiling := True;
+          FKeymanDeveloperOptions.Write;
+        end;
+      end;
+
+      if not modActionsMain.actFileSave.Execute then Exit;
+    end
+    else
+      Exit;
```

Comment:
So when ProjecFile is modified and silent is true, this just exists as NOOP?

Reply:
Correct. It exits with a false return value so the caller knows that things didn't work out.